File: objective-c/tutorial2/tutorial2/ViewController.m
#import "ViewController.h"
@import RMQClient;

@interface ViewController ()
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self workerNamed:@"Jack"];
    [self workerNamed:@"Jill"];
    sleep(1);
    [self newTask:@"Hello World..."];
    [self newTask:@"Just one this time."];
    [self newTask:@"Five....."];
    [self newTask:@"None"];
    [self newTask:@"Two..dots"];
}

- (void)newTask:(NSString *)msg {
    RMQConnection *conn = [[RMQConnection alloc] initWithDelegate:[RMQConnectionDelegateLogger new]];
    [conn start];

    id<RMQChannel> ch = [conn createChannel];

    RMQQueue *q = [ch queue:@"task_queue" options:RMQQueueDeclareDurable];

    [ch.defaultExchange publish:msg routingKey:q.name];
    NSLog(@"Sent %@", msg);

    [conn close];
}

- (void)workerNamed:(NSString *)name {
    RMQConnection *conn = [[RMQConnection alloc] initWithDelegate:[RMQConnectionDelegateLogger new]];
    [conn start];

    id<RMQChannel> ch = [conn createChannel];

    RMQQueue *q = [ch queue:@"task_queue" options:RMQQueueDeclareDurable];

    [ch basicQos:@1 global:NO];
    NSLog(@"%@: Waiting for messages", name);

    [q subscribe:^(id<RMQMessage>  _Nonnull message) {
        NSLog(@"%@: Received %@", name, message.content);
        // imitate some work
        unsigned int sleepTime = (unsigned int)[message.content componentsSeparatedByString:@"."].count - 1;
        NSLog(@"%@: Sleeping for %u seconds", name, sleepTime);
        sleep(sleepTime);
    }];
}

- (void)workerWithManualAckNamed:(NSString *)name {
    RMQConnection *conn = [[RMQConnection alloc] initWithDelegate:[RMQConnectionDelegateLogger new]];
    [conn start];

    id<RMQChannel> ch = [conn createChannel];

    RMQQueue *q = [ch queue:@"task_queue" options:RMQQueueDeclareDurable];

    [ch basicQos:@1 global:NO];
    NSLog(@"%@: Waiting for messages", name);

    RMQBasicConsumeOptions manualAck = RMQBasicConsumeNoOptions;
    [q subscribe:manualAck handler:^(id<RMQMessage>  _Nonnull message) {
        NSLog(@"%@: Received %@", name, message.content);
        // imitate some work
        unsigned int sleepTime = (unsigned int)[message.content componentsSeparatedByString:@"."].count - 1;
        NSLog(@"%@: Sleeping for %u seconds", name, sleepTime);
        sleep(sleepTime);

        [ch ack:message.deliveryTag];
    }];
}

@end

File: swift/tutorial2/tutorial2/ViewController.swift
//
//  ViewController.swift
//  tutorial2
//
//  Copyright © 2016 RabbitMQ. All rights reserved.
//
import UIKit
import RMQClient

class ViewController: UIViewController {


    override func viewDidLoad() {
        super.viewDidLoad()
        self.workerNamed("Jack")
        self.workerNamed("Jill")
        sleep(1)
        self.newTask("Hello World...")
        self.newTask("Just one this time.")
        self.newTask("Five.....")
        self.newTask("None")
        self.newTask("Two..dots")
    }

    func newTask(_ msg: String) {
        let conn = RMQConnection(delegate: RMQConnectionDelegateLogger())
        conn.start()
        let ch = conn.createChannel()
        let q = ch.queue("task_queue", options: .durable)
        let msgData = msg.data(using: .utf8)
        ch.defaultExchange().publish(msgData, routingKey: q.name, persistent: true)
        print("Sent \(msg)")
        conn.close()
    }

    func workerNamed(_ name: String) {
        let conn = RMQConnection(delegate: RMQConnectionDelegateLogger())
        conn.start()
        let ch = conn.createChannel()
        let q = ch.queue("task_queue", options: .durable)
        ch.basicQos(1, global: false)
        print("\(name): Waiting for messages")
        let manualAck = RMQBasicConsumeOptions()
        q.subscribe(manualAck, handler: {(_ message: RMQMessage) -> Void in
            let messageText = String(data: message.body, encoding: .utf8)
            print("\(name): Received \(messageText!)")
            // imitate some work
            let sleepTime = UInt32(messageText!.components(separatedBy: ".").count) - 1
            print("\(name): Sleeping for \(sleepTime) seconds")
            sleep(sleepTime)
            ch.ack(message.deliveryTag)
        })
    }
}
